File: THCS/2023_2024/VinhPhuc_2324/Bai3/spiralp.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

// Cấu trúc để lưu thông tin ô trên lưới
struct Cell {
    int x, y; // tọa độ
};

// Hàm kiểm tra xem ô có hợp lệ không
bool isValid(int x, int y, int N, int M, vector<vector<int>>& grid) {
    return (x >= 0 && x < N && y >= 0 && y < M && grid[x][y] == 0);
}

// Hàm tìm số bước tối thiểu để đến ô (N, M)
int bfsMinSteps(int N, int M, int K, vector<vector<int>>& grid) {
    queue<Cell> q;
    vector<vector<bool>> visited(N, vector<bool>(M, false));
    q.push({0, 0});
    visited[0][0] = true;

    int steps = 0;

    while (!q.empty()) {
        int size = q.size();
        for (int i = 0; i < size; ++i) {
            Cell cell = q.front();
            q.pop();

            // Kiểm tra nếu đã đến đích
            if (cell.x == N - 1 && cell.y == M - 1) {
                return steps;
            }

            // Thực hiện di chuyển từ ô hiện tại
            int currentIndex = cell.x * M + cell.y + 1; // Đánh số ô bắt đầu từ 1
            for (int move = 1; move <= K; ++move) {
                for (int j = 0; j <= move; ++j) {
                    int newX = cell.x + (move - j);  // Di chuyển xuống
                    int newY = cell.y + j;            // Di chuyển sang phải

                    if (isValid(newX, newY, N, M, grid) && !visited[newX][newY]) {
                        visited[newX][newY] = true;
                        q.push({newX, newY});
                    }
                }
            }
        }
        steps++;
    }

    return -1; // Nếu không đến được ô (N, M)
}

// Hàm tính số cách di chuyển hợp lệ
void calculateWays(int N, int M, int K, vector<vector<int>>& grid, vector<int>& ways) {
    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < M; ++j) {
            if (grid[i][j] == 0) { // Nếu ô là trắng
                int index = i * M + j + 1;
                for (int move = 1; move <= K; ++move) {
                    for (int jMove = 1; jMove <= move; ++jMove) {
                        int newX = i + (move - jMove);
                        int newY = j + jMove;

                        if (isValid(newX, newY, N, M, grid)) {
                            ways[index]++;
                        }
                    }
                }
            }
        }
    }
}

int main() {
    // Đọc dữ liệu
    int N, M, K;
    cin >> N >> M >> K;

    vector<vector<int>> grid(N, vector<int>(M));
    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < M; ++j) {
            cin >> grid[i][j];
        }
    }

    // Tính số bước tối thiểu
    int minSteps = bfsMinSteps(N, M, K, grid);

    // Tính số cách di chuyển hợp lệ
    vector<int> ways(N * M + 1, 0);
    calculateWays(N, M, K, grid, ways);

    // Tìm giá trị max(F(1), F(2), ..., F(N*M))
    int maxWays = *max_element(ways.begin() + 1, ways.end());

    // In kết quả
    cout << minSteps << " " << maxWays << endl;

    return 0;
}

File: THCS/2023_2024/VinhPhuc_2324/Bai3/spiralp.py
from collections import deque

def is_valid(x, y, N, M, grid):
    return 0 <= x < N and 0 <= y < M and grid[x][y] == 0

def bfs_min_steps(N, M, K, grid):
    queue = deque([(0, 0)])  # Bắt đầu từ ô (0, 0)
    visited = [[False] * M for _ in range(N)]
    visited[0][0] = True
    steps = 0

    while queue:
        size = len(queue)
        for _ in range(size):
            x, y = queue.popleft()

            # Kiểm tra nếu đã đến ô (N-1, M-1)
            if (x, y) == (N - 1, M - 1):
                return steps

            current_index = x * M + y + 1  # Đánh số ô bắt đầu từ 1
            for move in range(1, K + 1):
                for j in range(move + 1):
                    new_x = x + (move - j)  # Di chuyển xuống
                    new_y = y + j            # Di chuyển sang phải

                    if is_valid(new_x, new_y, N, M, grid) and not visited[new_x][new_y]:
                        visited[new_x][new_y] = True
                        queue.append((new_x, new_y))
        steps += 1

    return -1  # Nếu không đến được ô (N-1, M-1)

def calculate_ways(N, M, K, grid):
    ways = [0] * (N * M + 1)
    for i in range(N):
        for j in range(M):
            if grid[i][j] == 0:  # Nếu ô là trắng
                index = i * M + j + 1
                for move in range(1, K + 1):
                    for jMove in range(move + 1):
                        new_x = i + (move - jMove)
                        new_y = j + jMove

                        if is_valid(new_x, new_y, N, M, grid):
                            ways[index] += 1
    return ways

def main():
    # Đọc dữ liệu từ file 'spiralp.inp'
    with open('spiralp.inp', 'r') as f:
        N, M, K = map(int, f.readline().split())
        grid = [list(map(int, f.readline().split())) for _ in range(N)]

    # Tính số bước tối thiểu
    min_steps = bfs_min_steps(N, M, K, grid)

    # Tính số cách di chuyển hợp lệ
    ways = calculate_ways(N, M, K, grid)

    # Tìm giá trị max(F(1), F(2), ..., F(N*M))
    max_ways = max(ways[1:])  # Bỏ qua cách số 0

    # Ghi kết quả vào file 'spiralp.out'
    with open('spiralp.out', 'w') as out_file:
        out_file.write(f"{min_steps} {max_ways}\n")

if __name__ == "__main__":
    main()
